lib/puppet/agent.rb

```diff
@@ -1,6 +1,167 @@
 require 'puppet/application'
 require 'puppet/error'
 
+# A class providing pre/post fork callbacks modeled after those registered
+# by the `pthread_atfork` function.
+# The callbacks defined in this class are no-ops on all platforms except
+# Solaris where they ensure the forked process runs in a different contract
+# than the parent process. This is necessary in order for the child process
+# (an agent run) to be able to restart/stop the puppet agent service without
+# being killed (as a result of the termination of the parent process
+# contract during the service restart/stop).
+# The implementation uses Fiddle to make calls to functions defined in the
+# libcontract library.
+class AtFork
+  require 'fiddle'
+
+  raise LoadError, %q(The loaded fiddle version doesn't define 'Fiddle::Handle') unless defined?(Fiddle::Handle)
```

Comment:
Ah, sorry I get it now... One concern is that if we are running on Solaris, and we fail to map one of the functions (maybe a newer solaris version changes a method signature?), then we raise a LoadError, catch it, and silently "downgrade" to noop methods. So we could regress on this issue and not know it.

Could we instead define two implementations, one for Solaris, and one for everyone else, something like we do for FileSystem:

```ruby
class AtFork
  @impl = if Facter.value(:operatingsystem) == 'Solaris'
            SolarisAtFork.new
          else
             NoopAtFork.new
          end

  def prepare
    @impl.prepare
  end
end

class NoopAtFork
  def prepare
  end
  ...
end

class SolarisAtFork
  begin
    libhandle = Fiddle::Handle.new(library)
  rescue NameError => e
    raise _("The loaded fiddle version doesn't define 'Fiddle::Handle'")
  rescue Fiddle::DLError => e
    raise LoadError, e.message, e.backtrace
  end
  
  def prepare
    activate_new_contract_template
  end
end
```